### PR TITLE
Add Virtual Batch Normalization

### PR DESCRIPTION
**Pull request by saksham189**

### CODE REVIEW

**Comment by saksham189** [comment]
@ShikharJ  this should be ready for review.
I have implemented according to these implementations which are both consistent:
Pytorch: https://github.com/openai/improved-gan/blob/master/imagenet/model.py#L554
Tensorflow: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/gan/python/features/python/virtual_batchnorm_impl.py#L110

**Comment by saksham189** [comment]
@ShikharJ  this should be ready for review.
I have implemented according to the implementations provided in [Pytorch](https://github.com/openai/improved-gan/blob/master/imagenet/model.py#L554) and [Tensorflow](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/gan/python/features/python/virtual_batchnorm_impl.py#L110) which are both consistent with each other.

File: src/mlpack/methods/ann/layer/virtual_batch_norm.hpp [hunk 1]
```diff
@@ -0,0 +1,194 @@
+/**
+ * @file virtual_batch_norm.hpp
+ * @author Saksham Bansal
+ *
+ * Definition of the Virtual Batch Normalization layer class.
```
Comment (ShikharJ) [inline src/mlpack/methods/ann/layer/virtual_batch_norm.hpp]: Please change it to `VirtualBatchNorm` to stay consistent with `MiniBatchDiscrimination`.

File: src/mlpack/methods/ann/layer/virtual_batch_norm.hpp [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+#ifndef MLPACK_METHODS_ANN_LAYER_VIRTUALBATCHNORM_HPP
+#define MLPACK_METHODS_ANN_LAYER_VIRTUALBATCHNORM_HPP
+
+#include <mlpack/prereqs.hpp>
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+/**
+ * Declaration of the Virtual Batch Normalization layer class. Instead of using the
```
Comment (ShikharJ) [inline src/mlpack/methods/ann/layer/virtual_batch_norm.hpp]: Same here.

File: src/mlpack/methods/ann/layer/virtual_batch_norm.hpp [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+ * batch statistics for normalizing on a minibatch, it uses a reference subset of
```
Comment (ShikharJ) [inline src/mlpack/methods/ann/layer/virtual_batch_norm.hpp]: I think it should be `mini-batch` here.

File: src/mlpack/methods/ann/layer/virtual_batch_norm.hpp [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+class VirtualBatchNorm
+{
+ public:
+  //! Create the VirtualBatchNorm object.
+  VirtualBatchNorm();
+
+  /**
+   * Create the VirtualBatchNorm layer object for a specified number of input units.
+   *
+   * @param referenceBatch the data from which the normalization
```
Comment (ShikharJ) [inline src/mlpack/methods/ann/layer/virtual_batch_norm.hpp]: Capitalization here.

File: src/mlpack/methods/ann/layer/virtual_batch_norm.hpp [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+   * @param size The number of input units.
+   * @param eps The epsilon added to variance to ensure numerical stability.
+   */
+  template<typename eT>
+  VirtualBatchNorm(const arma::Mat<eT>& referenceBatch,
+                   const size_t size,
+                   const double eps = 1e-8);
+
+  /**
+   * Reset the layer parameters
```
Comment (ShikharJ) [inline src/mlpack/methods/ann/layer/virtual_batch_norm.hpp]: Punctuation at the end of sentence.

File: src/mlpack/methods/ann/layer/virtual_batch_norm.hpp [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+   */
+  void Reset();
+
+  /**
+   * Forward pass of the Virtual Batch Normalization layer. Transforms the input data
+   * into zero mean and unit variance, scales the data by a factor gamma and
+   * shifts it by beta.
+   *
+   * @param input Input data for the layer
```
Comment (ShikharJ) [inline src/mlpack/methods/ann/layer/virtual_batch_norm.hpp]: Punctuation at the end of sentence.

File: src/mlpack/methods/ann/layer/virtual_batch_norm.hpp [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+   * @param output Resulting output activations.
+   */
+  template<typename eT>
+  void Forward(const arma::Mat<eT>&& input, arma::Mat<eT>&& output);
+
+  /**
+   * Backward pass through the layer.
+   *
+   * @param input The input activations
```
Comment (ShikharJ) [inline src/mlpack/methods/ann/layer/virtual_batch_norm.hpp]: Same here.

File: src/mlpack/methods/ann/layer/virtual_batch_norm.hpp [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+  template<typename eT>
+  void Backward(const arma::Mat<eT>&& /* input */,
+                arma::Mat<eT>&& gy,
+                arma::Mat<eT>&& g);
+
+  /**
+   * Calculate the gradient using the output delta and the input activations.
+   *
+   * @param input The input activations
+   * @param error The calculated error
```
Comment (ShikharJ) [inline src/mlpack/methods/ann/layer/virtual_batch_norm.hpp]: In the above two statements here.

File: src/mlpack/methods/ann/layer/virtual_batch_norm.hpp [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+  OutputDataType const& Delta() const { return delta; }
+  //! Modify the delta.
+  OutputDataType& Delta() { return delta; }
+
+  //! Get the gradient.
+  OutputDataType const& Gradient() const { return gradient; }
+  //! Modify the gradient.
+  OutputDataType& Gradient() { return gradient; }
+
+
```
Comment (ShikharJ) [inline src/mlpack/methods/ann/layer/virtual_batch_norm.hpp]: Please remove the extra line here.

File: src/mlpack/methods/ann/layer/virtual_batch_norm.hpp [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+  /**
+   * Serialize the layer
```
Comment (ShikharJ) [inline src/mlpack/methods/ann/layer/virtual_batch_norm.hpp]: Punctuation at the end of sentence please.

File: src/mlpack/methods/ann/layer/virtual_batch_norm.hpp [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+  void serialize(Archive& ar, const unsigned int /* version */);
+
+ private:
+  //! Locally-stored number of input units.
+  size_t size;
+
+  //! Locally-stored epsilon value.
+  double eps;
+
+  //! Whether we are in loading or saving mode.
```
Comment (ShikharJ) [inline src/mlpack/methods/ann/layer/virtual_batch_norm.hpp]: I think this can be something like `Variable to keep track of whether we are in ...`

File: src/mlpack/methods/ann/layer/virtual_batch_norm.hpp [hunk 12]
```diff
@@ -0,0 +1,194 @@
...
+  //! Locally-stored output parameter object.
+  OutputDataType outputParameter;
+
+  //! Locally-stored input parameter object.
+  OutputDataType inputParameter;
+
+  //! Locally-stored normalized input.
+  OutputDataType normalized;
+
+  //! Locally-stored 0 mean input.
```
Comment (ShikharJ) [inline src/mlpack/methods/ann/layer/virtual_batch_norm.hpp]: Please make it `zero mean input`.

File: src/mlpack/methods/ann/layer/virtual_batch_norm_impl.hpp [hunk 13]
```diff
@@ -0,0 +1,152 @@
+/**
+ * @file virtual_batch_norm_impl.hpp
+ * @author Saksham Bansal
+ *
+ * Implementation of the Virtual Batch Normalization Layer.
```
Comment (ShikharJ) [inline src/mlpack/methods/ann/layer/virtual_batch_norm_impl.hpp]: Same here. Also `layer` should not be capitalized.

File: src/mlpack/methods/ann/layer/virtual_batch_norm_impl.hpp [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+    const double eps) :
+    size(size),
+    eps(eps),
+    loading(false)
+{
+  weights.set_size(size + size, 1);
+
+  referenceBatchMean = arma::mean(referenceBatch, 1);
+  referenceBatchMeanSquared = arma::mean(arma::square(referenceBatch), 1);
+  newCoefficient = 1.0/(referenceBatch.n_cols + 1);
```
Comment (ShikharJ) [inline src/mlpack/methods/ann/layer/virtual_batch_norm_impl.hpp]: Please add a couple of spaces around the divide sign.

File: src/mlpack/methods/ann/layer/virtual_batch_norm_impl.hpp [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void VirtualBatchNorm<InputDataType, OutputDataType>::Backward(
+    const arma::Mat<eT>&& /* input */, arma::Mat<eT>&& gy, arma::Mat<eT>&& g)
+{
+  const arma::mat stdInv = 1.0 / arma::sqrt(variance + eps);
+
+  // Step 1: dl / dxhat
```
Comment (ShikharJ) [inline src/mlpack/methods/ann/layer/virtual_batch_norm_impl.hpp]: I guess we can make use of punctuation here and also in the `BatchNorm` and `LayerNorm` code.

File: src/mlpack/methods/ann/layer/virtual_batch_norm.hpp (hunk 11 above)
Comment (saksham189) [reply src/mlpack/methods/ann/layer/virtual_batch_norm.hpp]: we are using the same description in `LayerNorm` and `BatchNorm`

Comment (ShikharJ) [reply src/mlpack/methods/ann/layer/virtual_batch_norm.hpp]: Then they should be changed as well :)

File: src/mlpack/methods/ann/layer/virtual_batch_norm_impl.hpp [hunk 16]
```diff
@@ -0,0 +1,148 @@
+/**
+ * @file virtual_batch_norm_impl.hpp
+ * @author Saksham Bansal
+ *
+ * Implementation of the Virtual Batch Normalization layer.
```
Comment (ShikharJ) [inline src/mlpack/methods/ann/layer/virtual_batch_norm_impl.hpp]: Please change this to `VirtualBatchNorm`.

File: src/mlpack/methods/ann/layer/virtual_batch_norm_impl.hpp [hunk 17]
```diff
@@ -0,0 +1,148 @@
...
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void VirtualBatchNorm<InputDataType, OutputDataType>::Backward(
+    const arma::Mat<eT>&& /* input */, arma::Mat<eT>&& gy, arma::Mat<eT>&& g)
+{
+  const arma::mat stdInv = 1.0 / arma::sqrt(variance + eps);
+
+  // Step 1: dl / dxhat.
```
Comment (ShikharJ) [inline src/mlpack/methods/ann/layer/virtual_batch_norm_impl.hpp]: Could you please ad this punctuation to `BatchNorm` and `LayerNorm` code as well?

File: src/mlpack/methods/ann/layer/virtual_batch_norm.hpp [hunk 18]
```diff
@@ -0,0 +1,193 @@
...
+  void serialize(Archive& ar, const unsigned int /* version */);
+
+ private:
+  //! Locally-stored number of input units.
+  size_t size;
+
+  //! Locally-stored epsilon value.
+  double eps;
+
+  //! Variable to keep track of whether we are in loading or saving mode.
```
Comment (ShikharJ) [inline src/mlpack/methods/ann/layer/virtual_batch_norm.hpp]: Same changes in `BatchNorm` and `LayerNorm` please.

**Comment by ShikharJ** [comment]
@saksham189 Can you update this? This seems ready to be merged in my opinion.

File: src/mlpack/methods/ann/layer/virtual_batch_norm_impl.hpp [hunk 19]
```diff
@@ -0,0 +1,148 @@
...
+
+  // Step 2: sum dl / dxhat * (x - mu) * -0.5 * stdInv^3.
+  const arma::mat var = arma::sum(norm % temp, 1) %
+      arma::pow(stdInv, 3.0) * -0.5;
+
+  g = (norm.each_col() % stdInv) + ((inputParameter.each_col() %
+      var) * 2 * newCoefficient / inputParameter.n_cols);
+
+  g.each_col() += (arma::sum(norm.each_col() % -stdInv, 1) + (var %
+      mean * -2)) * newCoefficient / inputParameter.n_cols;
```
Comment (ShikharJ) [inline src/mlpack/methods/ann/layer/virtual_batch_norm_impl.hpp]: Some descriptive comments for the above two lines would be helpful for people wishing to get a quick explanation. Just like the ones from `BatchNorm`.

Comment (saksham189) [reply src/mlpack/methods/ann/layer/virtual_batch_norm_impl.hpp]: I have removed `Step 1` , `Step 2` etc. as in `BatchNorm` because that actually refers to the derivations of backprop given in the `BatchNorm` paper.

**Comment by ShikharJ** [comment]
Hmm, my bad, I didn't realize `inputMean` was already a local variable name, maybe try changing to `inputSubMean`?

**Comment by ShikharJ** [comment]
Thanks @saksham189 for the nice contribution.
